src/Raven.Client/Documents/BulkInsert/BulkInsertOperation.cs

```diff
@@ -423,14 +517,40 @@ private async Task HandleErrors(string documentId, Exception e)
             await ThrowOnUnavailableStream(documentId, e).ConfigureAwait(false);
         }
 
-        private IDisposable ConcurrencyCheck()
+        private IDisposable _streamLockDisposable;
+        private  ValueTask<ReleaseStream> ConcurrencyCheckAsync()
         {
             if (Interlocked.CompareExchange(ref _concurrentCheck, 1, 0) == 1)
                 throw new InvalidOperationException("Bulk Insert store methods cannot be executed concurrently.");
 
-            return new DisposableAction(() => Interlocked.CompareExchange(ref _concurrentCheck, 0, 1));
+            if (_streamLock.Wait(0))
```

Comment:
not sure I understand the mechanism here, first we are checking if we can take a lock (and taking it if possible) and if we can't then we are waiting for the lock? shouldn't we just wait for the lock?

Reply:
The idea here is that if we can get the lock in an uncontended manner, we can skip the `Task` allocation and return immediately.
We'll only allocate the `Task` if we couldn't acquire the lock, so we wait.

I'm assuming that usually we won't have a contention, and that saves us an allocation per call

Reply:
Ok, makes sense